### PR TITLE
Move AsyncQueue to spin locks

### PR DESCRIPTION
**Pull request by brianpepin**
This PR primarily moves async queue away from std::mutex and instead uses spin locks, at least on multi-processor systems. The reason for this is to prevent context switches while adding things to the queue, which can slow down perf critical threads.

The async queue implementation is also now built on an interface abstraction, which will make interoperability between Windows code and libHttpClient easier.  

Finally, there are some API additions / changes:

1. DuplicateAsyncQueueHandle now returns a HRESULT and has the new handle as an out-param.  The semantics haven't changed, but this signature mates better with other Win32 "duplicate" APIs and allows us a little more freedom in the implementation later on.

2. There are two new dispatch modes:  SerializedThreadPool uses the thread pool, but ensures that queue calls are run in series through it instead of in parallel. This is useful if you want to use a thread pool as your completion dispatch mode, but still want things to always come in order.  Immediate dispatch mode doesn't queue at all and just invokes the callback immediately on the submitting thread.  This is a useful utility mechanism for chaining async calls together.

3. RemoveAsyncQueueCallbacks can now accept a nullptr as the callback function pointer to act as a wild card.  This must be used with caution, since you can remove any callback (not just yours), but it's kind of necessary to support lambdas.

### CODE REVIEW

File: Source/Task/asyncQueueP.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    AsyncQueueSection,
+    AsyncQueue
+};
+
+// Lightweight version of IUnknown
+struct IApi
+{
+    virtual uint32_t __stdcall AddRef() = 0;
+    virtual uint32_t __stdcall Release() = 0;
+    virtual HRESULT __stdcall QueryApi(_In_ ApiId id, _Out_ void** ptr) = 0;
```
Comment (lubeltra) [inline Source/Task/asyncQueueP.h]: do we really need this?
QueryInterface is subtle and annoying to implement, do we really need runtime type information in this code?

File: Source/Task/AsyncQueueEx.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+#include "AsyncQueueImpl.h"
+
+static uint64_t const INVALID_SHARE_ID = 0xFFFFFFFFFFFFFFFF;
+
+static LIST_ENTRY s_sharedList;
+static spinlock s_sharedLock;
+static std::once_flag s_sharedOnceFlag;
+
+#define MAKE_SHARED_ID(threadId, workMode, completionMode) \
+    (uint64_t)(((uint64_t)workMode) << 48 | ((uint64_t)completionMode) << 32 | threadId)
```
[the remark this answers is not in the document]
Comment (brianpepin) [reply Source/Task/AsyncQueueEx.cpp]: "threadId" is actually an outdated parameter name. This is the ID you pass to MakeSharedAsyncQueue.  It's defined as a uint32_t and it's *common* to use the thread ID, but not mandatory.

File: Source/Task/AsyncQueueEx.cpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    {
+        aq = new (std::nothrow) AsyncQueueEx;
+        RETURN_IF_NULL_ALLOC(aq);
+        RETURN_IF_FAILED(aq->Initialize(workDispatchMode, completionDispatchMode));
+
+        std::unique_lock<spinlock> lock(s_sharedLock);
+        referenced_ptr<AsyncQueueEx> aq2 = AsyncQueueEx::FindSharedQueue(queueId);
+        if (aq2 != nullptr)
+        {
+            // We raced with someone else
```
[the remark this answers is not in the document]
Comment (brianpepin) [reply Source/Task/AsyncQueueEx.cpp]: It shouldn't - it's a referenced_ptr so it releases on destruction.

[the remark this answers is not in the document]
Comment (lubeltra) [reply Source/Task/AsyncQueueEx.cpp]: ah yes, I missed the declaration above

**Pull request merged**